src/kem/kyber/serialize.rs

```diff
@@ -41,29 +42,47 @@ use super::{
 /// is called, and `compress_q` also performs this conversion.
 
 #[inline(always)]
-fn serialize_little_endian_1<const OUT_LEN: usize>(
+pub(crate) fn compress_then_serialize_message(
     re: KyberPolynomialRingElement,
-) -> [u8; OUT_LEN] {
-    let mut serialized = [0u8; OUT_LEN];
+) -> [u8; SHARED_SECRET_SIZE] {
+    let mut serialized = [0u8; SHARED_SECRET_SIZE];
 
-    for (i, chunk) in re.coefficients.chunks_exact(8).enumerate() {
-        for (j, coefficient) in chunk.iter().enumerate() {
-            serialized[i] |= (*coefficient as u8) << j
+    for (i, coefficients) in re.coefficients.chunks_exact(8).enumerate() {
+        for (j, coefficient) in coefficients.iter().enumerate() {
+            let coefficient = to_unsigned_representative(*coefficient);
+
+            let coefficient_compressed = compress_q::<1>(coefficient);
+            debug_assert!(coefficient_compressed == 0 || coefficient_compressed == 1);
+
+            // At this point, the following should hold for |coefficient_compressed|:
+            //
+            // if coefficient_compressed == 0 {
+            //     coefficient < 832 ||
+            //     (coefficient >= 2496 && coefficient < 4161)
+            // } else { // coefficient_compressed = 1
+            //     (coefficient >= 832 && coefficient < 2496) ||
+            //     (coefficient >= 4161 && coefficient < 5825)
+            // }
+            //
+            // TODO(xvzcf): When this is turned into an assertion, intermittent
```

Comment:
The ciphertext in that test is probably not a valid ciphertext. Hence it would make sense that these sometimes fail.
So you could add a feature flag or something to disable these for those tests. Or not consider these asserts failures in those tests and retry if they panic here.